### PR TITLE
Support the MoE model in RLinf 

### PR DESCRIPTION
**Pull request by FxxxxU**
### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: rlinf/hybrid_engines/sglang/sglang_0_4_6/sgl_scheduler.py [hunk 1]
```diff
@@ -198,8 +206,31 @@ def sync_hf_weight(self, recv_req: SyncHFWeightInput):
...
+                batch_weights,
+                disable=self.get_parent_rank() != 0 or self.tp_rank != 0,
+                desc="Load weights",
+            ):
+                model.load_weights(bucket)
+                for name, weight in bucket:
+                    del weight
+                bucket.clear()
+            torch.cuda.empty_cache()
         else:
```
Comment (andylin-hao) [inline rlinf/hybrid_engines/sglang/sglang_0_4_6/sgl_scheduler.py]: What about disaggregate mode?

File: rlinf/hybrid_engines/sglang/sglang_0_4_6/sgl_scheduler.py [hunk 2]
```diff
@@ -41,6 +41,7 @@
     suppress_other_loggers,
 )
```
Comment (andylin-hao) [inline rlinf/hybrid_engines/sglang/sglang_0_4_6/sgl_scheduler.py]: What about other sglang versions?

File: rlinf/utils/resharding/mcore_weight_reshard.py [hunk 3]
```diff
@@ -158,8 +192,61 @@ def get_layer_num(param_name):
...
+        if reshard_ep_model:
+            # if use the te group gemm, we need to convert the weight type from te group to seq group
+            if self.config.moe_grouped_gemm == "te":
+                from toolkits.ckpt_convertor.utils.mg_moe_groupgemm import (
+                    moe_te_group_to_seq,
+                )
+
+                expert_params = moe_te_group_to_seq(expert_params)
+            else:
+                assert self.config.moe_grouped_gemm == "legacy", (
```
Comment (qurakchin) [inline rlinf/utils/resharding/mcore_weight_reshard.py]: the assert equation is different from the message.

Comment (FxxxxU) [reply rlinf/utils/resharding/mcore_weight_reshard.py]: fix

File: rlinf/hybrid_engines/sglang/sglang_0_4_6/sgl_scheduler.py (hunk 2 above)
Comment (FxxxxU) [reply rlinf/hybrid_engines/sglang/sglang_0_4_6/sgl_scheduler.py]: fix sglang version 0_4_4, 0_4_6, 0_4_9

File: rlinf/hybrid_engines/sglang/sglang_0_4_6/sgl_scheduler.py (hunk 1 above)
Comment (FxxxxU) [reply rlinf/hybrid_engines/sglang/sglang_0_4_6/sgl_scheduler.py]: update the disaggregate mode

File: toolkits/ckpt_convertor/convert_middle_file_to_mg.py [hunk 4]
```diff
@@ -133,13 +134,13 @@ def gen_checkpoint_name(pp_rank, ep_rank_input):
             else:
                 ep_rank = (model_rank - tpe_rank) // target_tpe
 
-            if target_tpe * target_ep > target_tp:
+            # fix: for Megatron 0.11.0, patch the megatron load ckpt from huggingface
+            # if target_tpe * target_ep <= target_tp:
```
Comment (guozhen1997) [inline toolkits/ckpt_convertor/convert_middle_file_to_mg.py]: Please explain the reason for this modification.

Comment (FxxxxU) [reply toolkits/ckpt_convertor/convert_middle_file_to_mg.py]: fix

File: rlinf/utils/resharding/mcore_weight_reshard.py [hunk 5]
```diff
@@ -158,8 +196,61 @@ def get_layer_num(param_name):
...
+        if reshard_ep_model:
+            # if use the te group gemm, we need to convert the weight type from te group to seq group
+            if self.config.moe_grouped_gemm == "te":
+                from toolkits.ckpt_convertor.utils.mg_moe_groupgemm import (
+                    moe_te_group_to_seq,
+                )
+
+                expert_params = moe_te_group_to_seq(expert_params)
+            else:
+                assert self.config.moe_grouped_gemm in [None, "te"], (
```
Comment (Lin-xs) [inline rlinf/utils/resharding/mcore_weight_reshard.py]: ”te“ has been handled in previous if statement.

Comment (FxxxxU) [reply rlinf/utils/resharding/mcore_weight_reshard.py]: This is to reserve some interfaces for subsequent groupgemm.

File: rlinf/utils/resharding/utils.py [hunk 6]
```diff
@@ -69,6 +83,83 @@ def tp_reshard_fn_qwen2_5(model_state_dict, merge_factor, tp_group):
...
+        if (
+            "rotary_pos_emb.inv_freq" in k
+            or "linear_qkv.layer_norm_weight" in k
+            or "linear_fc1.layer_norm_weight" in k
+            or "final_layernorm.weight" in k
+            or "q_layernorm.weight" in k
+            or "k_layernorm.weight" in k
+            or "pre_mlp_layernorm.weight" in k
+            or "router.weight" in k
+        ):
```
Comment (Lin-xs) [inline rlinf/utils/resharding/utils.py]:
```python
param_skip_tp_reshard = ["rotary_pos_emb.inv_freq", ....]
if any(name in k for name in param_skip_reshard):
    ...
```

File: rlinf/utils/resharding/utils.py [hunk 7]
```diff
@@ -95,8 +186,40 @@ def _gather_pp_group_tensor_and_reshard(
...
+    )
+    if tensor is not None:
+        model_state_dict[key] = tensor.clone()
+
+    key = "output_layer.weight"
+    tensor = _gather_pp_group_tensor_and_reshard(
+        model_state_dict, key, pp_last_rank, pp_group, dtype
+    )
+    if tensor is not None:
+        model_state_dict[key] = tensor.clone()
```
Comment (Lin-xs) [inline rlinf/utils/resharding/utils.py]: Use a for loop to simplify these codes.

**Review by andylin-hao — COMMENTED** [review]
Please add ci for this

File: toolkits/ckpt_convertor/utils/mg_moe_groupgemm.py [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+import torch
+
+
+def moe_seq_to_te_group(state_dict_):
```
Comment (Lin-xs) [inline toolkits/ckpt_convertor/utils/mg_moe_groupgemm.py]: Why should we add an underline as suffix in `state_dict_`?

File: toolkits/ckpt_convertor/convert_middle_file_to_mg.py [hunk 9]
```diff
@@ -610,6 +615,17 @@ def convert_layer(
...
+    if convert_config.grouped_gemm == "te":
+        for model_key in saver.full_checkpoint:
+            state_dict = saver.full_checkpoint[model_key]["model"]
+            moe_seq_to_te_group(state_dict)
+    elif convert_config.grouped_gemm == "legacy":
+        for model_key in saver.full_checkpoint:
+            state_dict = saver.full_checkpoint[model_key]["model"]
+            moe_seq_to_group(state_dict, num_local_experts, glu=True)
+    elif convert_config.grouped_gemm is not None:
+        assert False
```
Comment (Lin-xs) [inline toolkits/ckpt_convertor/convert_middle_file_to_mg.py]: Please add error message here.

File: rlinf/config.py [hunk 10]
```diff
@@ -492,6 +510,20 @@ def validate_megatron_cfg(cfg: DictConfig) -> DictConfig:
         cfg.model.expert_model_parallel_size = cfg.model.expert_model_parallel_size = (
             cfg.model.get("expert_model_parallel_size", 1)
         )
+        cfg.model.expert_tensor_parallel_size = (
+            cfg.model.expert_tensor_parallel_size
+        ) = cfg.model.get("expert_tensor_parallel_size", 1)
```
Comment (Lin-xs) [inline rlinf/config.py]: Is this assignment chain necessary? @guozhen1997 
If not, we can remove it.

File: rlinf/config.py [hunk 11]
```diff
@@ -1001,8 +1034,15 @@ def build_transformer_config(cfg) -> "TransformerConfig":
...
         "tp_only_amax_red": tp_only_amax_red,
+        "qk_layernorm": cfg.get("qk_layernorm", False),
+        "kv_channels": cfg.get("head_dim", None),
         # MoE related
         "num_moe_experts": cfg.get("num_moe_experts", None),
+        "moe_ffn_hidden_size": cfg.get("moe_ffn_hidden_size", None),
+        # now the sequential mlp should ffn hidden size == moe_ffn_hidden_size
+        "ffn_hidden_size": cfg.get("moe_ffn_hidden_size", None)
+        if cfg.get("moe_ffn_hidden_size", None)
+        else cfg.get("ffn_hidden_size", None),
```
Comment (Lin-xs) [inline rlinf/config.py]: Can be simplified by
```python
"ffn_hidden_size": cfg.get("moe_ffn_hidden_size", None) or cfg.get("ffn_hidden_size", None)
```
By the way, please assert `ffn_hidden_size == moe_ffn_hidden_size` and add error message if it is necessary.

File: rlinf/utils/resharding/utils.py [hunk 12]
```diff
@@ -95,8 +186,40 @@ def _gather_pp_group_tensor_and_reshard(
...
+    return model_state_dict
+
 
+def pp_reshard_fn_qwen3_moe(model_state_dict, pp_group, dtype):
```
Comment (qurakchin) [inline rlinf/utils/resharding/utils.py]: It seems totally same to pp_reshard_fn_qwen2_5.

File: examples/reasoning/config/math/qwen3-30b-grpo-megatron.yaml [hunk 13]
```diff
@@ -0,0 +1,304 @@
...
+
+    # MoE related
+    moe_router_load_balancing_type: "aux_loss"
+    init_method_std: 0.008
+    moe_router_topk: 2
+    moe_router_pre_softmax: False
+    moe_aux_loss_coeff: 0.0
+
+    # qwen3 related
+    qk_layernorm: True
```
Comment (qurakchin) [inline examples/reasoning/config/math/qwen3-30b-grpo-megatron.yaml]: you can move this config into validate_model_cfg_by_hf_config

File: rlinf/config.py (hunk 10 above)
Comment (FxxxxU) [reply rlinf/config.py]: fix

File: rlinf/config.py (hunk 11 above)
Comment (FxxxxU) [reply rlinf/config.py]: can‘t add assert, because qwen3 ffn_hidden_size != moe_ffn_hidden_size in qwen3 hf config.

File: examples/reasoning/config/math/qwen3-30b-grpo-megatron.yaml (hunk 13 above)
Comment (FxxxxU) [reply examples/reasoning/config/math/qwen3-30b-grpo-megatron.yaml]: done

File: rlinf/utils/resharding/utils.py (hunk 12 above)
Comment (FxxxxU) [reply rlinf/utils/resharding/utils.py]: update

File: toolkits/ckpt_convertor/convert_middle_file_to_mg.py (hunk 9 above)
Comment (FxxxxU) [reply toolkits/ckpt_convertor/convert_middle_file_to_mg.py]: done

File: toolkits/ckpt_convertor/utils/mg_moe_groupgemm.py (hunk 8 above)
Comment (FxxxxU) [reply toolkits/ckpt_convertor/utils/mg_moe_groupgemm.py]: update

File: rlinf/utils/resharding/utils.py (hunk 7 above)
Comment (FxxxxU) [reply rlinf/utils/resharding/utils.py]: done

File: rlinf/utils/resharding/utils.py (hunk 6 above)
Comment (FxxxxU) [reply rlinf/utils/resharding/utils.py]: done

File: rlinf/utils/resharding/utils.py [hunk 14]
```diff
@@ -50,22 +64,149 @@ def _gather_tp_group_tensor_and_reshard(tensor, dim, merge_factor, tp_group):
...
 def tp_reshard_fn_qwen2_5(model_state_dict, merge_factor, tp_group):
+    # Parameters that should skip TP resharding (just clone)
+    param_skip_tp_reshard = [
+        "linear_qkv.layer_norm_weight",
+        "mlp.linear_fc1.layer_norm_weight",
+        "final_layernorm.weight",
+    ]
+
+    # Parameters that need to be gathered on dim=0
+    param_reshard_column_parallel_linear = [
```
Comment (Lin-xs) [inline rlinf/utils/resharding/utils.py:75]: Parameters that need to be gathered on dim=0 are column_parallel or row_parallel?

File: rlinf/utils/resharding/utils.py [hunk 15]
```diff
@@ -50,22 +64,149 @@ def _gather_tp_group_tensor_and_reshard(tensor, dim, merge_factor, tp_group):
 
 
 def tp_reshard_fn_qwen2_5(model_state_dict, merge_factor, tp_group):
+    # Parameters that should skip TP resharding (just clone)
+    param_skip_tp_reshard = [
+        "linear_qkv.layer_norm_weight",
+        "mlp.linear_fc1.layer_norm_weight",
+        "final_layernorm.weight",
+    ]
```
Comment (Lin-xs) [inline rlinf/utils/resharding/utils.py:72]: what about `rotary_pos_emb.inv_freq`?

File: rlinf/utils/resharding/utils.py [hunk 16]
```diff
@@ -94,37 +235,41 @@ def _gather_pp_group_tensor_and_reshard(
...
-        model_state_dict, key, pp_last_rank, pp_group, dtype
-    )
-    if tensor is not None:
-        model_state_dict[key] = tensor.clone()
+    keys_with_ranks = [
+        ("decoder.final_layernorm.weight", pp_last_rank),
+        ("decoder.final_layernorm.bias", pp_last_rank),
+        ("embedding.word_embeddings.weight", pp_first_rank),
+        ("output_layer.weight", pp_last_rank),
+    ]
```
Comment (Lin-xs) [inline rlinf/utils/resharding/utils.py:261]: pp_first_rank first?

File: examples/reasoning/config/math/qwen3-30b-grpo-megatron.yaml [hunk 17]
```diff
@@ -0,0 +1,301 @@
...
+  - override hydra/job_logging: stdout
+
+hydra:
+  run:
+    dir: .
+  output_subdir: null
+
+cluster:
+  num_nodes: 2
+  num_gpus_per_node: 8
```
Comment (andylin-hao) [inline examples/reasoning/config/math/qwen3-30b-grpo-megatron.yaml]: Remove this

Comment (FxxxxU) [reply examples/reasoning/config/math/qwen3-30b-grpo-megatron.yaml]: fix
